File: q3346/q3346.py
class Solution:
    def maxFrequency(self, nums: List[int], k: int, numOperations: int) -> int:
        nums.sort()
        counter = Counter(nums)
        maxFreq = 0
        # use each number in the nums array as a target to converge on
        for target in range(nums[0], nums[-1] + 1):
            # min(all possible numbers, numOperations + existing numbers)
            currFreq = min(self.getRange(nums, target, k), 
                           counter[target] + numOperations)
            # update the max frequency while enumerating
            maxFreq = max(maxFreq, currFreq)
        return maxFreq
    
    # total number in the nums array that can be made into target by adding [-k, k]
    def getRange(self, nums, target, k):
        left = bisect_left(nums, target - k)
        right = bisect_right(nums, target + k)
        return right - left

File: q3346/q3346.java
public class Solution {
    public int maxFrequency(int[] nums, int k, int numOperations) {
        Arrays.sort(nums);
        Map<Integer, Integer> counter = new HashMap<>();
        for (int num : nums) {
            counter.put(num, counter.getOrDefault(num, 0) + 1);
        }

        int maxFreq = 0;
        int minVal = nums[0];
        int maxVal = nums[nums.length - 1];

        for (int target = minVal; target <= maxVal; target++) {
            int rangeCount = getRange(nums, target, k);
            int existing = counter.getOrDefault(target, 0);
            int currFreq = Math.min(rangeCount, existing + numOperations);
            maxFreq = Math.max(maxFreq, currFreq);
        }

        return maxFreq;
    }

    // Helper: counts how many numbers are within [target - k, target + k]
    private int getRange(int[] nums, int target, int k) {
        int left = lowerBound(nums, target - k);
        int right = upperBound(nums, target + k);
        return right - left;
    }

    // Equivalent of Python's bisect_left
    private int lowerBound(int[] nums, int target) {
        int lo = 0, hi = nums.length;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] < target) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        return lo;
    }

    // Equivalent of Python's bisect_right
    private int upperBound(int[] nums, int target) {
        int lo = 0, hi = nums.length;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (nums[mid] <= target) {
                lo = mid + 1;
            } else {
                hi = mid;
            }
        }
        return lo;
    }
}

